src/semantic/type_checker.rs

```diff
@@ -0,0 +1,161 @@
+#![allow(unused_variables)]
+use lexical::types::Types;
+use syntax::expr::ExprWrapper;
+use syntax::expr::Expr::*;
+use semantic::analyzer_trait::ASTAnalyzer;
+
+pub struct TypeChecker;
+
+impl TypeChecker {
+    pub fn new() -> TypeChecker {
+        TypeChecker
+    }
+
+    fn cmp_lhs_rhs(lhs_str: String, rhs_str: String) -> Option<String> {
```

Comment:
really? that's sad. But still, the rest applies still i think? It should be recursively returning the types. to coerce towards the lhs what the type is

And you don't think we should be sending the types, themselves and encapsulating what you do with the type (e.g. checking the type name if its custom) in here?